File: pa1/Lex.java

/* JULIAN TO (jcto)
 * PA 1
 * CS 101
 * PROFESSOR TANTALO
 * SPRING 2019
 */

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.Scanner;

public class Lex {
	public static void main(String[] args) throws IOException {

		// System.out.println("begin program");

		// checks for valid input files
		if (args.length < 2) {
			System.err.println("ERROR: BAD INPUT");
			System.exit(1);
		}

		// read in input file
		Scanner in = new Scanner(new File(args[0]));
		PrintWriter out = new PrintWriter(new FileWriter(args[1]));

		// counts number of lines in the input file
		int counter = 0;
		while (in.hasNextLine()) {
			counter++;
			in.nextLine();
		}

		in.close();
		in = null;
		String[] input = new String[counter];
		// open file again
		in = new Scanner(new File(args[0]));

		int koala = 0;
		while (in.hasNextLine()) {
			input[koala] = in.nextLine();
			koala++;
		}
		//System.out.println(counter + " strings read");
		for (int x = 0; x < counter; x++) {
			//System.out.println(input[x]);
		}

		// insertion sort
		List poi = new List();
		//start the list 
		poi.append(0);
		int i, j = 0;
		for (i = 1; i < counter; i++) {
			j = i - 1;
			String temp = input[i];
			poi.moveBack();
			//move the element backwards on the list
			//until it is lexicographically less than the one in front
			while (j >= 0 && temp.compareTo(input[poi.get()]) <= 0) {
				poi.movePrev();
				j--;
			}
			if (poi.index() >= 0) {
				poi.insertAfter(i);
			} else {
				poi.prepend(i);
			}
		}
		poi.moveFront();
		//System.out.println(poi);
		//System.out.println(input[poi.get()]);
		//System.out.println(poi.length());
		
		//write output to file
		for(int f = 0; f < poi.length(); f++)
		{
			//System.out.println(input[poi.get()]);
			out.println(input[poi.get()]);
			poi.moveNext();
		}
		// close files
		in.close();
		out.close();
	}
}

File: pa2/Lex.c
/*
JULIAN TO (jcto)
pa2: CMPS 101
Professor Tantalo
Spring 2019
*/
// Based upon the provided FILE.IO
// and logic in Lex.java

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "List.h"

#define MAX_LEN 160

int main(int argc, char *argv[]) {

  int n, count = 0;
  FILE *in, *out;
  char line[MAX_LEN];
  // char tokenlist[MAX_LEN];
  // char *token;

  // check command line for correct number of arguments
  if (argc != 3) {
    printf("Usage: %s <input file> <output file>\n", argv[0]);
    exit(1);
  }

  // open files for reading and writing
  in = fopen(argv[1], "r");
  out = fopen(argv[2], "w");
  if (in == NULL) {
    printf("Unable to open file %s for reading\n", argv[1]);
    exit(1);
  }
  if (out == NULL) {
    printf("Unable to open file %s for writing\n", argv[2]);
    exit(1);
  }

  // first count num of lines in the files
  while (fgets(line, MAX_LEN, in) != NULL) {
    count++;
  }
  // holds all of the lines of file
  // 2d array: 1st entry is line num, 2nd is length of word
  char Word[count][MAX_LEN];

  // go back to the top of the FILE
  rewind(in);
  // copies inputs of file into Word array
  while (fgets(line, MAX_LEN, in) != NULL) {
    strcpy(Word[n], line);
    n++;
  }
  fclose(in);

  // make a new LIST
  List listing = newList();
  // list starts at zero, obviously
  append(listing, 0);

  // same logic as Lex.java really
  int i = 0;
  for (i = 1; i < count; i++) {
    moveFront(listing);

    while (index(listing) >= 0) {
      char *temp = Word[i];

      // move back until you reach a word lexicographically less
      if (strncmp(Word[get(listing)], temp, MAX_LEN) > 0) {
        insertBefore(listing, i);
        break;
      }
      moveNext(listing);
    }
    // if you reach the front of the list, stop there
    if (index(listing) < 0) {
      append(listing, i);
    }
  }

  moveFront(listing);
  // print the list
  // for testing
  int track = 0;
  printf("value of count is %d\n", count);
  for (track = 0; track < count; track++) {
    printf("%s on itration: %d\n", Word[get(listing)], track);
    fprintf(out, "%s", Word[get(listing)]);
    moveNext(listing);
    // track = track + 1;
  }

  printf("leave the loop?");
  /* close files */
  fclose(in);
  fclose(out);
  // free list
  freeList(&listing);
  printf("free the list?");

  return (0);
}
